src/core/labeling/qgspallabeling.cpp

```diff
@@ -483,6 +483,25 @@ bool QgsPalLayerSettings::prepare( QgsRenderContext &context, QSet<QString> &att
   return true;
 }
 
+QSet<QString> QgsPalLayerSettings::referencedFields() const
+{
+  QSet<QString> referenced;
+  if ( drawLabels )
+  {
+    if ( isExpression )
+    {
+      referenced.unite( QgsExpression( fieldName ).referencedColumns() );
+    }
+    else
+    {
+      referenced.insert( fieldName );
+    }
+  }
+
+  referenced.unite( mDataDefinedProperties.referencedFields( QgsExpressionContext(), true ) );
+  return referenced;
```

Comment:
There's also the text format and callout's data defined properties to consider

Reply:
Ah right, I must have missed callout's DDP and geometry generator's referenced fields too.

Regarding text format - I can't find text format's DDP considered anywhere in QgsPalLayerSettings::prepare() - shouldn't it be there too?

Reply:
Ah sorry, the text format data defined properties aren't used in labeling, they are only used elsewhere. But callouts should still be considered